### DISCUSSION TITLE
OpenSSL RSA key pair generation not working with BIO object on MAC

### DISCUSSION THREAD

**Discussion by aniljadaun** [openssl/openssl#21558] [Q&A]
HI Team

Below code is not working for generating public key using openssl 3.1.1

        // 2. save public key
        BIO* pub = BIO_new(BIO_s_mem());
        i2d_PUBKEY_bio(pub, pkey);
        int pub_len = BIO_pending(pub);
        ACUns8 pub_key[8192];
        BIO_read(pub, pub_key, pub_len);

However this works

        BIO* pub = BIO_new(BIO_s_mem());
        int pub_len = BIO_pending(pub);
        ACUns8 pub_key[8192];
        ACUns8* pptr = &pub_key[0];
        pub_len = i2d_PUBKEY(pkey, NULL);
        i2d_PUBKEY(pkey,&pptr);
        
  Is bio_read deprecated in openssl 3.1.1 or is this a bug ? please suggest?

**Comment by mattcaswell** [comment]
What exactly is not working? Do you get an error return value from any of the calls above? What is the value of `pub_len` after the `BIO-pending()` call?

**Reply by aniljadaun** [reply]
@mattcaswell  No error is thrown as such. 
pub_len after BIO-pending() is zero .

**Reply by mattcaswell** [reply]
So, is the i2d_PUBKEY_bio() call successful? From the man page:

````
i2d_PUBKEY_fp() and i2d_PUBKEY_bio() return 1 if successfully
encoded or 0 if an error occurs.
````
So is the return value from that function 1 or 0? If 0 you might want to check if there are any errors reported on the OpenSSL error stack, e.g. by inserting `ERR_print_errors_fp(stderr)` after the `i2d_PUBKEY_bio()` call.

**Reply by aniljadaun** [reply]
Hi @mattcaswell 

Thanks for your suggestion. 

i2d_PUBKEY_bio returns 0 which means it is not successful. Error reported is as below

00B66A0202000000:error:0308010C:digital envelope routines:inner_evp_generic_fetch:unsupported:crypto/evp/evp_fetch.c:341:Global default library context, Algorithm (mdc2 : 98), Properties (<null>)
00B66A0202000000:error:0308010C:digital envelope routines:inner_evp_generic_fetch:unsupported:crypto/evp/evp_fetch.c:341:Global default library context, Algorithm (mdc2 : 98), Properties (<null>)
00B66A0202000000:error:0308010C:digital envelope routines:inner_evp_generic_fetch:unsupported:crypto/evp/evp_fetch.c:341:Global default library context, Algorithm (mdc2 : 98), Properties (<null>)

**Reply by aniljadaun** [reply]
Hi @mattcaswell 

Can you please suggest what could be reason behind this ? Am i doing something wrong in this ? What does this errror means ?

**Reply by mattcaswell** [reply]
Just to rule out he possibility that this is an old stale error on the error stack please could you insert a call to `ERR_clear_errors()` immediately before the call to `i2d_PUBKEY_bio()` and confirm that you still see the same error messages?

**Reply by aniljadaun** [reply]
ERR_clear_errors() doing this no error is dumped...

**Reply by aniljadaun** [reply]
Hi @mattcaswell 

I am not getting any errors in ERR_print_errors_fp though the call is not successful. Can you please check if this is bug at your end ?

**Comment by slontis** [comment]
I just tried running this code by simply adding your code to a function that is run by demos/pkey/EVP_PKEY_RSA_keygen and adding the call at the end of generate_rsa_key_long() and it works just fine.. (pub_len is 550 bytes). 
I suspect that your PKEY is not valid.
You may want to try doing the same thing to see if it works for you (follow the instructions in the demos/pkey/makefile to run).

**Reply by aniljadaun** [reply]
Hi @slontis 

Thanks for looking into this. Could you please elaborate on this ? PKEY is not valid ?

This code is working fine for my workflow
    BIO* pub = BIO_new(BIO_s_mem());
    int pub_len = BIO_pending(pub);
    ACUns8 pub_key[8192];
    ACUns8* pptr = &pub_key[0];
    pub_len = i2d_PUBKEY(pkey, NULL);
    i2d_PUBKEY(pkey,&pptr);

**Comment by slontis** [comment]
Try running the demo - and see if it works.. It does a keygen, add your code after the keygen call and see if it works..

**Reply by aniljadaun** [reply]
// The default value for key length
		keyLength =2048;

		EVP_PKEY_CTX* ctx = NULL;
		EVP_PKEY* pkey = NULL;
		ctx = EVP_PKEY_CTX_new_id(EVP_PKEY_RSA, NULL);

		bool success = true;
		if (EVP_PKEY_keygen_init(ctx) <= 0) {
			success = false;
		}

		if (success && EVP_PKEY_CTX_set_rsa_keygen_bits(ctx, keyLength) <= 0) {
			success = false;
		}

		if (success && EVP_PKEY_keygen(ctx, &pkey) <= 0) {
			success = false;
		}
		
		This is key generation logic i am using

**Reply by aniljadaun** [reply]
@slontis 
Code used in our workflow for generating rsakey is different from the demo code shared by you .Please refer my previous comment. Do you see any issue with our version , it is old code . we would not want to change the old logic if not necessary. 
Please suggest.

**Reply by slontis** [reply]
That also works for me by changing the demo to use code similar to yours.... You may have to debug it, to figure out where it goes wrong.
On my ubuntu machine this code works on both master and on the 3.1.1 tar source.

**Reply by aniljadaun** [reply]
@slontis  This is working for me in sample project but not working in my App code environment :( . Can i use

BIO* pub = BIO_new(BIO_s_mem());
int pub_len = BIO_pending(pub);
ACUns8 pub_key[8192];
ACUns8* pptr = &pub_key[0];
pub_len = i2d_PUBKEY(pkey, NULL);
i2d_PUBKEY(pkey,&pptr);

Do you see any issues with this code ?

**Reply by aniljadaun** [reply]
@slontis Also i tried dumping key pair generated in my app and it worked fine for me. 


Number of bits: 2048

Public values:
  n = 0xD189032C1FC09B7406F0FA4F13C20259CBAFE7DBDB48C39ED40709E8311D819D0D06FB7FD480AB1C4A61E3D4914CD93A6300759939DCB0F5C5A95517DCB6D6F9A67DC471020B07F41C769FF295C8BC5C61AD5B80FBBC9265FC2270C647494B31A8D4AE93C4712E20855C2A0873411C684FF1C63509C5B4D74032A0B819894006C949607CF0CAC97AC381ADA7720D3FD163711E5AA591AE4F229A36146D7D77FAE87F3E3F3072C543FBF46F08884F812C1F98F1C87816CB6EF4E8D0B1874DA4E1E08BC9E00F193EB137479E0AF49B0231CD79BBA664D93D8C7157938B15B656FFD5166413B86E2E2C07BF9BECAEA1E7F56DDED4DC25A2893661DE993E340C81B3
  e = 0x10001


-----BEGIN PUBLIC KEY-----
MIIBIjANBgkqhkiG9w0BAQEFAAOCAQ8AMIIBCgKCAQEA0YkDLB/Am3QG8PpPE8IC
Wcuv59vbSMOe1AcJ6DEdgZ0NBvt/1ICrHEph49SRTNk6YwB1mTncsPXFqVUX3LbW
+aZ9xHECCwf0HHaf8pXIvFxhrVuA+7ySZfwicMZHSUsxqNSuk8RxLiCFXCoIc0Ec
aE/xxjUJxbTXQDKguBmJQAbJSWB88MrJesOBradyDT/RY3EeWqWRrk8imjYUbX13
+uh/Pj8wcsVD+/RvCIhPgSwfmPHIeBbLbvTo0LGHTaTh4IvJ4A8ZPrE3R54K9JsC
Mc15u6Zk2T2McVeTixW2Vv/VFmQTuG4uLAe/m+yuoef1bd7U3CWiiTZh3pk+NAyB
swIDAQAB
-----END PUBLIC KEY-----

**Reply by aniljadaun** [reply]
Does this means pkey is actually valid , but some how  below code is not working. 

 // 2. save public key
    BIO* pub = BIO_new(BIO_s_mem());
    i2d_PUBKEY_bio(pub, pkey);
    int pub_len = BIO_pending(pub);
    ACUns8 pub_key[8192];
    BIO_read(pub, pub_key, pub_len);

**Comment by slontis** [comment]
The question you should be asking yourself is why does similar code not work in your app. (You will need to figure that out yourself).

**Reply by aniljadaun** [reply]
@slontis Thanks for your help. On debugging the workflow i found first call of creating memory on bio fails.

 BIO *pub = BIO_new(BIO_s_mem());
    if (pub == NULL) {
        printf("Failed to create memory BIO\n");
        return -1;
    }
    
    
  Do you have any suggestion on what could be the reason for this. ?

**Comment by slontis** [comment]
Not really.. (You are not reusing the pub variable or something silly hopefully).

It must be failing in BIO_new_ex() somewhere so you would need to debug that also, to see where it falls over.. If it is inside the method->create()..

Also, Maybe you are trashing memory somewhere?? valgrind might help find that..
